### DISCUSSION TITLE
Pitfalls to using synchronous and async engine in same app?

### DISCUSSION THREAD

**Discussion by Shawn-Struble** [sqlalchemy/sqlalchemy#10344] [Usage Questions] [answered]
Hi,

I'm working on an application that makes use of synchronous SQLAlchemy and has many different jobs running. I would like to introduce SQLAlchemy asyncio into some of the jobs while leaving the synchronous code in place for the vast majority of jobs that would be unchanged.

I am aware of features like: [run_sync](https://docs.sqlalchemy.org/en/20/orm/extensions/asyncio.html#sqlalchemy.ext.asyncio.AsyncConnection.run_sync) for running otherwise synchronous code in an asyncio application. 

My question is if it's okay to instantiate and use two engines of different drivers in the same app:
```python
sync_db_url = "postgres://user:pass@localhost/test"  # psycopg2
async_db_url = "postgresql+asyncpg://user:pass@localhost/test"
sync_engine = create_engine(sync_db_url)
async_engine = create_async_engine(async_db_url)
```

Is this possible or ill-advised? I can't seem to find anywhere this is talked about explicitly in any detail.

Cheers,
Shawn

#### SQLAlchemy Version in Use
2.0.20
#### DBAPI 
psycopg2
asyncpg
#### Database Vendor and Major Version
PostgreSQL 12+
#### Python Version
3.10
#### Operating system
Linux

**Comment by wangxin688** [comment]
yes. it works for me. I use both async and sync in the same fastapi project with two diffrent session depends

**Comment by CaselIT** [comment, marked as answer]
Hi,

The main drawback is probably that you cannot share a transaction between the two.
If that's not a problem then I don't see major things. Just consider that the two engine are independent, so the number of connections to postgresql can be higher with the default configuration, since each engine has its own pool.

If you use psycopg version 3 you can also use the same dbapi on both engines.

> I am aware of features like: [run_sync](https://docs.sqlalchemy.org/en/20/orm/extensions/asyncio.html#sqlalchemy.ext.asyncio.AsyncConnection.run_sync) for running otherwise synchronous code in an asyncio application.

Please note that you can run synchronous sqlalchemy code only. If you try to use other blocking features, like time.sleep() or request, you will block the asyncio loop

**Reply by Shawn-Struble** [reply]
Thank you!

**Reply by jgourinda** [reply]
Come out to the discussion while trying to achieve the same thing. 

But in my use case I need to be able to share a transaction between two engines sync and async. 

The whole app is running in async mode but since we needed to introduce Celery which does not support async code as of now we had to add a sync engine.  

For our tests we use the great pattern shared in [Joining a Session into an External Transaction (such as for test suites)](https://docs.sqlalchemy.org/en/20/orm/session_transaction.html#joining-a-session-into-an-external-transaction-such-as-for-test-suites). But the issue is that the celery worker does not see the data written to db in async test transaction. 

Before introducing Celery we had the below fixture that took care of giving an async Session for any test that needs it and it also makes the mainland `MainAsyncSession` used by app code join the test transaction. 

```py
@pytest_asyncio.fixture(autouse=True)
async def Session(xdist_worker_id: str) -> t.AsyncIterator[async_sessionmaker]:
    async with get_db_engine(tenant=xdist_worker_id).connect() as connection:
        await connection.begin()
        try:
            _Session = async_sessionmaker(
                bind=connection,
                join_transaction_mode="create_savepoint",
                expire_on_commit=False,
                autoflush=True,
            )
            async_bind = MainAsyncSession.kw["bind"]
            MainAsyncSession.configure(bind=connection)
            yield _Session
        except Exception:
            raise
        finally:
            MainAsyncSession.configure(bind=async_bind)
            await connection.rollback()
```

Of course trying to do `Main**Sync**Session.configure(bind=connection)` in the code above does not work. 

You did say 
> The main drawback is probably that you cannot share a transaction between the two. 

But it was some time ago so I don't know if it might be possible now. 

I don't really know where the transaction state lies. At the Engine level or at the DB level so that it might be able to be joined by the Sync Engine via some sort of internal id ?

Anyway thank for the great material and amount of information available in the docs and the repo.

**Reply by CaselIT** [reply]
No, it's still not possible.

Your best option is likely run_sync to execute sqlalchemy come in a sync manner. But care must be taken to avoid using other sync features that may block the loop, like sleep or requests.

**Reply by jgourinda** [reply]
Thank you !

**Comment by shahriar-mohim007** [comment]
How were you able to inherit both AsyncBase and SyncBase in the same model?

**Reply by CaselIT** [reply]
I'm not sure what those are. They are not sqlalchemy classes

**Reply by shahriar-mohim007** [reply]
AsyncBase = declarative_base(cls=AsyncAttrs)
SyncBase = declarative_base()


class Base(AsyncBase, SyncBase):
        pass                                           
is it possible to achieve ?I want to use same model for synchronous and async operation

**Reply by CaselIT** [reply]
You only need one of those, don't create two bases
